Templatize helm template image references

joestringer opened the pull request:
From the main commit:

    Move all of the declarations of the image repositories and tags into
    install/kubernetes/Makefile.values, then templatize the values file so
    that we can pick up these env variables using envsubst. Then we can get
    rid of the big ugly 'sed' hack in the Makefile which is hard to maintain.

Broadly this PR moves the `install/kubernetes/cilium/values.yaml` to `install/kubernetes/cilium/values.yaml.tmpl` as the source of truth. This new values file template has variables wherever the docker repositories and tags are.

Inside the Makefiles, we declare the repos/tags as variables, then when we run `envsubst` against the `install/kubernetes/cilium/values.yaml.tmpl`, it substitutes the variables into the template and generates the final output `install/kubernetes/cilium/values.yaml` file.

This looks like a big change (~2KLoC), but most of it is just making a copy of `install/kubernetes/values.yaml` in order to templatize it. Review commit-by-commit to see incremental changes.

This PR also fixes the default helm in the master branch to point towards `quay.io/cilium/cilium-ci:latest` so that you can install directly from the repo. When we branch for the next release, we will update the repo and tag to point to the branch instead and then the helm charts should work for that target development branch.

You can also override these default settings by making your own copy of `install/kubernetes/Makefile.values` and then running this with `MAKEFILE_VALUES=/path/to/my/values.mk make -C install/kubernetes`.

joestringer committed:
Makefile.digests: Update filepath reference

The autogenerated file referred to a workflow that doesn't exist, update
how that path is generated to reflect the current location.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
Makefile.digests: Export all variables

Export each of the digest variables so that they can be passed to
subcommands from future Makefile infrastructure.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
install/kubernetes: Update generated files

Now that the generation code has been overhauled in prior commits,
update the generated files based on the latest code.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
install/kubernetes: Split Chart.yaml update target

Split the target used for updating Chart.yaml, since the substitutions
there are a bit simpler and unrelated to the rest of the changes we need
to do for the values.yaml in an upcoming commit.

This introduces a new install/kubernetes/Makefile.values file with an
exported variable for the cilium branch, used for generating a logo URL
for external websites like https://artifacthub.io .

While we're at it, drop the non-cross-platform `-Z` for separating paths
by NULL character to parse the inputs to iterate them and run commands
against them.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
install: Templatize image repositories and tags

Move all of the declarations of the image repositories and tags into
install/kubernetes/Makefile.values, then templatize the values file so
that we can pick up these env variables using envsubst. Then we can get
rid of the big ugly 'sed' hack in the Makefile which is hard to maintain.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
install/kubernetes: Default to installing latest

Default these helm templates to install the latest version from the tip
of master, for easy development.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
install: Reuse exported vars in check-docker-images

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer commented on install/kubernetes/Makefile.digests:
```diff
@@ -1,13 +1,13 @@
-OPERATOR_ALIBABACLOUD_DIGEST := ""
-OPERATOR_GENERIC_DIGEST := ""
-OPERATOR_DIGEST := ""
+export CILIUM_DIGEST := ""
```
The alternative to all this exporting is to just blanket export all variables to subshells via `.EXPORT_ALL_VARIABLES` as a [Special Target](https://www.gnu.org/software/make/manual/html_node/Special-Targets.html), but I wasn't sure if that might cause other issues since I don't know if that's scoped only to this Makefile or whether it could end up exporting other vars.

michi-covalent commented on install/kubernetes/Makefile:
```diff
@@ -100,14 +57,7 @@ lint:
+check-docker-images: # Check whether docker images are available for the current version.
+	$(QUIET)$(ROOT_DIR)/contrib/release/check-docker-images.sh "$(CILIUM_VERSION)"
+
+.PHONY: all check-docker-images clean docs lint update-chart update-versions
```
does `cilium/values.yaml` need to be added as a phony target, or is it enough to just have `update-versions` here?

joestringer replied:
Every target that is *not* a file should be there.

michi-covalent reviewed (approved):
lgtm just one question about phony target. this is much cleaner 🚢

aditighag commented on install/kubernetes/Makefile.values:
```diff
@@ -8,11 +8,11 @@ export USE_DIGESTS ?= $(shell if grep -q '""' $(DIGESTS_PATH); then echo "false"
 export PULL_POLICY:=Always# -> IfNotPresent
 export CILIUM_BRANCH:=master# -> vX.Y
-export CILIUM_REPO:=quay.io/cilium/cilium# -> quay.io/cilium/cilium
+export CILIUM_REPO:=quay.io/cilium/cilium-ci# -> quay.io/cilium/cilium
```
What does the "->" do?

joestringer replied:
Nothing, it's just a comment to remind me whenever we do the branch about what it should look like.

One unusual thing here is that I deliberately put the `#` immediately after the value to avoid extraneous whitespace showing up in variable names inside the charts.

Maybe this would be easier to get right during branch time if I dropped the trailing comments.

joestringer replied:
^ Changed the format for this, I realized that I'll probably get it wrong otherwise.

aditighag commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -0,0 +1,2041 @@
+# File generated by install/kubernetes/Makefile; DO NOT EDIT.
+# This file is based on install/kubernetes/cilium/values.yaml.tmpl.
```
How was this file generated? Helm?

joestringer replied:
It's generated using `envsubst`. Maybe I could have described this better in the PR description. Broadly what this PR does is it adds a bunch of exported variables in `install/kubernetes/Makefile.values`, then runs envsubst which takes in the `values.yaml.tmpl` on stdin plus the environment variables, then substitutes all of the variables in the template and writes out to this file.

joestringer replied:
^ I expanded the PR description to help explain this, hopefully that makes sense now.

aditighag reviewed (commented):
Thanks for tidying up the Makefiles, and fixing the defaults with "-ci" image tags.

joestringer commented on install/kubernetes/Makefile:
```diff
@@ -36,58 +32,19 @@ LOGO_PATH := Documentation/images/logo-solo.svg
+	$(ECHO_GEN)$@
+	echo -e "# File generated by $(RELATIVE_DIR)/Makefile; DO NOT EDIT." > $@
+	echo -e "# This file is based on $(RELATIVE_DIR)/cilium/values.yaml.tmpl.\n" >> $@
+	envsubst < $< >> $@
```
This is the magic line, basically take `$<` (the template file, first dependency in the list) as stdin to `envsubst`, which also takes the exported environment variables as inputs, then redirect the output to `$@` (the target `values.yaml` file).

joestringer committed:
docs: Document new values step when branching

Signed-off-by: Joe Stringer <joe@cilium.io>

kaworu commented on install/kubernetes/Makefile:
```diff
@@ -1,29 +1,25 @@
 
-CILIUM_CHARTS := "$(ROOT_DIR)/$(RELATIVE_DIR)/cilium"
+CILIUM_CHARTS := $(ROOT_DIR)/$(RELATIVE_DIR)/cilium
 CILIUM_VALUES := "$(CILIUM_CHARTS)/values.yaml"
```
`CILIUM_VALUES` seems unused after this patch. It is tempting to define something like:

```makefile
CILIUM_VALUES_TEMPLATE := $(CILIUM_VALUES).tmpl
```
And use both variable, but I would lean towards removing `CILIUM_VALUES` instead because the values files are only used by this Makefile (i.e. the relative path is enough) and the former solution make `git grep values.yaml.tmpl` worse.

kaworu commented on install/kubernetes/Makefile:
```diff
@@ -36,58 +32,19 @@ LOGO_PATH := Documentation/images/logo-solo.svg
+		| xargs -L 1 $(SED) -i -e 's/'$(CILIUM_CHART_REGEX)'/\1 '$(VERSION)'/g'
+	$(QUIET)$(SED) -i 's;icon:.*;icon: $(LOGO_BASE_URL)/cilium@$(CILIUM_BRANCH)/$(LOGO_PATH);' $(CHART_FILE)
+
+cilium/values.yaml: $(CILIUM_CHARTS)/values.yaml.tmpl $(ROOT_DIR)/VERSION $(ROOT_DIR)/$(RELATIVE_DIR)/Makefile.digests $(ROOT_DIR)/$(RELATIVE_DIR)/$(MAKEFILE_VALUES)
```
Now that `cilium/values.yaml` is generated, it should be added as a dependency to the `lint` target correct?

joestringer replied:
Good catches, thanks, will fix these up.

joestringer replied:
Ah this one's a bit interesting because lint should really depend on every single file inside install/kubernetes/cilium. I guess a simple find should do the trick.

kaworu reviewed (changes requested):
Thanks @joestringer! Couple of comments but overall LGTM.

On the CI side, `test-smoke.yaml` should detect changes to `values.yaml.tmpl` without a corresponding `values.yaml` update:

https://github.com/cilium/cilium/blob/9580ea3508ea694091a8a52321784597506de59a/.github/workflows/tests-smoke.yaml#L96-L99

Looking at this workflow, "Run helm-docs" seems extraneous because "Run helm-charts" execute `make -C install/kubernetes` and the default `all` target has the `docs` target as dependency. As this is kind of related to this PR, would you mind adding a commit to fix this (either remove "Run helm-docs" or make "Run helm-charts" execute the `update-versions` target)? Happy to open another PR if you prefer.

joestringer committed:
install/kubernetes: Tidy up lint target

Split up the variables used here for better readability, and add the V=0
quiet line to indicate progress in the way the other targets do.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
tests-smoke: Remove redundant helm smoke tests

The 'docs' target under install/kubernetes is already run by the
'helm-charts' step via the 'all' make target. Given some of the recent
re-organizing, we should probably just add linting to the overall steps
in the makefile and run all of the targets.

This should avoid running these targets twice:
* 2x Run cilium/values.yaml generation (via lint, all targets)
* 2x Run docs (via docs, all targets)

Suggested-by: Alexandre Perrin <alex@isovalent.com>
Signed-off-by: Joe Stringer <joe@cilium.io>

pchaigno reviewed (approved):
I have only reviewed the changes falling under my codeowners, but LGTM.

kaworu commented on install/kubernetes/Makefile:
```diff
@@ -34,80 +29,37 @@ HELM_DOCS := $(DOCKER_RUN) $(HELM_DOCS_IMAGE)
+ALL_HELM_FILES := $(shell find cilium/ -type f)
+CRD_FILES := $(shell find $(ROOT_DIR)/examples/crds/*/ -type f)
+CRDS := $(foreach path,$(patsubst %.yaml,%,$(CRD_FILES)),$(shell basename $(path)))
+lint: cilium/values.yaml $(CRD_FILES) $(ALL_HELM_FILES)
```
My mistake as I've previously asked to add `cilium/values.yaml` to the `lint` target, but since `lint` is phony it's going to be executed regardless of dependency updates.

```suggestion
CRD_FILES := $(shell find $(ROOT_DIR)/examples/crds/*/ -type f)
CRDS := $(foreach path,$(patsubst %.yaml,%,$(CRD_FILES)),$(shell basename $(path)))
lint:
```

joestringer replied:
I think `lint` still needs to depend on `cilium/values.yaml` to ensure that file has been generated against the latest template changes. I guess we can get rid of the extra find for `ALL_HELM_FILES` but it feels a bit weird to add only a subset of the dependencies into the target dependencies :thinking:

joestringer replied:
I guess it doesn't really matter, by default it'll generate that file and then run the linter. I'll take your suggestion here.

sayboras commented on install/kubernetes/Makefile:
```diff
@@ -34,80 +29,37 @@ HELM_DOCS := $(DOCKER_RUN) $(HELM_DOCS_IMAGE)
-         ../../contrib/release/check-docker-images.sh "v$(VERSION)"
-
-.PHONY: all check-docker-images clean docs lint update-versions
+check-docker-images: # Check whether docker images are available for the current version.
```
running this target in this PR somehow failed for existing images

```
docker.io/cilium/cilium-etcd-operator:v2.0.7@sha256:04b8327f7f992693c2cb483b999041ed8f92efc8e14f2a5f3ab95574a65ea2dc does not exist!
docker.io/cilium/startup-script:d69851597ea019af980891a4628fb36b7880ec26 does not exist!
...
```

joestringer replied:
Yeah, those images aren't on docker.io. Actually the target is only valid on older branches, it fails on master today:

```
$ make -C install/kubernetes/ check-docker-images
docker.io/cilium/startup-script:d69851597ea019af980891a4628fb36b7880ec26 does not exist!
docker.io/cilium/cilium:v1.11.90 does not exist!
quay.io/cilium/cilium:v1.11.90 does not exist!
docker.io/cilium/clustermesh-apiserver:v1.11.90 does not exist!
quay.io/cilium/clustermesh-apiserver:v1.11.90 does not exist!
docker.io/cilium/docker-plugin:v1.11.90 does not exist!
quay.io/cilium/docker-plugin:v1.11.90 does not exist!
docker.io/cilium/hubble-relay:v1.11.90 does not exist!
quay.io/cilium/hubble-relay:v1.11.90 does not exist!
docker.io/cilium/operator:v1.11.90 does not exist!
quay.io/cilium/operator:v1.11.90 does not exist!
docker.io/cilium/operator-azure:v1.11.90 does not exist!
quay.io/cilium/operator-azure:v1.11.90 does not exist!
docker.io/cilium/operator-aws:v1.11.90 does not exist!
quay.io/cilium/operator-aws:v1.11.90 does not exist!
docker.io/cilium/operator-generic:v1.11.90 does not exist!
quay.io/cilium/operator-generic:v1.11.90 does not exist!
make: *** [Makefile:104: check-docker-images] Error 1
```

So, net improvement. We can probably teach that script not to look up docker.io for those last two repos.

joestringer replied:
> Yeah, those images aren't on docker.io

That was actually a lie, it appears that the same digests are just not present on DockerHub. Maybe the more accurate statement would be, "we don't configure the charts to pull those images from DockerHub".

joestringer replied:
I added one more commit on top that should make this work on the master branch now.

sayboras reviewed (approved):
~I referred to values.yaml a lot, can we somehow keep the history of this file ? it's not a big deal, as I can check the history of values.yaml.tmpl as well.~ Updated: it was me relying on IDE like below comment.

One comment as per below, the rest looks legit :100:

joestringer committed:
contrib: Fix docker image check in tree

Previously this script wouldn't work when invoked on the master branch
via `make -C install/kubernetes check-docker-images`, because the script
expected some images to be present on dockerhub when we don't keep them
up to date there (startup-script, cilium-etcd-operator), and also
because the tree specifies the image tag as "latest" which are only
published on the quay.io/cilium/*-ci repositories. Teach the script to
pick up the correct set of image dependencies to handle these cases.

Signed-off-by: Joe Stringer <joe@cilium.io>

kaworu reviewed (approved):
Thanks @joestringer!
